File: db/entities/Product.ts
import {
  Entity,
  PrimaryKey,
  Property,
  ManyToOne,
  type Rel,
} from "@mikro-orm/core";
import { Vendor } from "./Vendor";

@Entity()
export class Product {
  @PrimaryKey({ type: "number" })
  id!: number;

  @Property({ type: "string" })
  name!: string;

  @Property({ type: "text", nullable: true })
  description?: string;

  @Property({ type: "string", nullable: true })
  category?: string; // e.g., "Clinical Module", "Analytics Platform", "API Service", etc.

  @Property({ type: "string", nullable: true })
  version?: string;

  @ManyToOne(() => Vendor)
  vendor!: Rel<Vendor>;

  @Property({ type: "Date" })
  createdAt: Date = new Date();

  @Property({ type: "Date", onUpdate: () => new Date() })
  updatedAt: Date = new Date();
}

File: api-python/app/models/product.py
from datetime import datetime

from sqlalchemy import DateTime, ForeignKey, Integer, String, Text
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.models.base import Base
from app.models.vendor import Vendor


class Product(Base):
    __tablename__ = "product"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    name: Mapped[str] = mapped_column(String(255), nullable=False)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    category: Mapped[str | None] = mapped_column(String(255), nullable=True)
    version: Mapped[str | None] = mapped_column(String(255), nullable=True)
    vendor_id: Mapped[int] = mapped_column(Integer, ForeignKey("vendor.id"), nullable=False)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)
    updated_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)

    vendor: Mapped["Vendor"] = relationship("Vendor", back_populates="products")
